File: runme/mcgen/local/datan/solsrc/s6lsq.f
      PROGRAM S6LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXNR=3,MAXN=100,T0=-3.D0)
      DIMENSION T(MAXN),DELTAY(MAXN),CX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DIMENSION Y(MAXN),X(MAXNR),SCRAT(MAXNR,MAXNR)
      CHARACTER*75 TX,TY,CAPT
      DIMENSION RANDOM(1000),HIST(100)
      DIMENSION XPL(100),YPL(100)
      DIMENSION LIST(MAXNR)
      EXTERNAL BWFNCT
      DATA NR /3/
C identify program to user
      WRITE(*,*)' Program S6LSQ generates data'
      WRITE(*,*)' corresponding to a Breit-Wigner law'
      WRITE(*,*)' and fits a Breit-Wigner to histogram'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number NEV (<= 1000)'
      WRITE(*,*)' of events to be generated'
      WRITE(*,*)' >'
      READ(*,*)NEV
      WRITE(*,*)' Enter number NT (<= 100) of histogram bins'
      WRITE(*,*)' >'
      READ(*,*)NT
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C generate events and build histogram
      CALL RNBW(0.D0,1.D0,RANDOM,NEV)
      DELT=2.D0*ABS(T0)/DBLE(NT)
      CALL SMHSIN(HIST,T0,DELT,NT)
      DO 5 I=1,NEV
        CALL SMHSFL(HIST,T0,DELT,NT,RANDOM(I),1.D0)
5     CONTINUE
      N=0
      DO 10 I=1,NT
        IF(HIST(I).GT.0.D0) THEN
          N=N+1
          T(N)=T0+(DBLE(I)-0.5D0)*DELT
          Y(N)=HIST(I)
          DELTAY(N)=SQRT(Y(N))
        END IF
10    CONTINUE
      IF(N.LE.NR) THEN
        WRITE(*,*)' only ',N,' bins with at least one event'
        STOP
      END IF
C set first approximation of unknowns
      X(1)=0.5D0
      X(2)=0.5D0
      X(3)=DBLE(NEV)
C perform fit
      NSTEP=100
      CALL LSQNON(BWFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,A,SCRAT,
     +NSTEP)
      P=1.D0-SCCHI2(R,N-NR)
      IF(NSTEP.LT.0) THEN
        WRITE(*,'(A,I5)')' LSQNON ended with NSTEP = ',NSTEP
      ELSE
C prepare graphics
        TX='t'
        TY='y'
        LTX=1
        LTY=1
        CAPT(1:33)='x_1#=*****,x_2=#*****,x_3#=*****,'
        CAPT(34:74)='&D@x_1#=*****,&D@x_2=#*****,&D@x_3#=*****'
        WRITE(CAPT(6:10),'(F5.2)')X(1)
        WRITE(CAPT(17:21),'(F5.2)')X(2)
        WRITE(CAPT(28:32),'(F5.1)')X(3)
        WRITE(CAPT(42:46),'(F5.2)')SQRT(CX(1,1))
        WRITE(CAPT(56:60),'(F5.2)')SQRT(CX(2,2))
        WRITE(CAPT(70:74),'(F5.1)')SQRT(CX(3,3))
        LCAPT=74
C curve corresponding to solution
        NPL=100
        DO 20 IPL=1,NPL
          XPL(IPL)=-4.D0+0.08D0*DBLE(IPL)
          YPL(IPL)=BWFNCT(X,NR,XPL(IPL))
20      CONTINUE
C graphical output
        CALL GRHSCV(XPL,YPL,NPL,HIST,T0,DELT,NT,TX,LTX,TY,LTY,CAPT,
     +  LCAPT,NWS)
      END IF
      END
C ------------------------------------------------------------------
      SUBROUTINE RNBW(A,GAMMA,R,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(PI=3.14159D0,HALF=0.5D0)
      DIMENSION R(N)
C program generates random numbers corresponding to
C a Breit-Wigner distribution
      CALL RNECUY(R,N)
      DO 10 I=1,N
        R(I)=A+HALF*GAMMA*TAN(PI*(R(I)-HALF))
10    CONTINUE
      END
C-------------------------------------------------
      DOUBLE PRECISION FUNCTION BWFNCT(X,NR,T)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(*)
      BWFNCT=X(3)*2.D0*X(2)**2/((3.14159*X(2))*(4.D0*(T-X(1))**2+X(2)
     +**2))
      END

File: runme/mcgen/local/datan/solsrc/s6lsq.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                         double P_T *r, integer const P_T *n);
extern double CBFUNC bwfnct(double const P_T *x, integer const P_T *nr,
                            double const P_T *t);

/* Table of constant values */

static double c_d0 = 0., c_d1 = 1., c_dm3 = -3.;

main(void)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */

    static integer i, n, ipl, nev, npl, lcapt, list[3], ltx, lty, nws,
                   nstep, nt, nr = 3;
    static double a[300] /* was [100][3] */, cx[9] /* was [3][3] */,
                  delt, hist[100], p, r, t[100], x[3], y[100], random[1000],
                  deltay[100], scrat[9] /* was [3][3] */, xpl[100], ypl[100];
    static char capt[75], tx[75], ty[75];

/* identify program to user */
    printf("%s\n", "Program S6LSQ generates data");
    printf("%s\n", "corresponding to a Breit-Wigner law");
    printf("%s\n\n", "and fits a Breit-Wigner to histogram.");
/* ask for input */
    printf("%s\n", "Enter number NEV (<= 1000)");
    printf("%s\n", "of events to be generated:");
    printf("%s", "> ");
    scanf("%li", &nev);
    printf("%s\n", "Enter number NT (<= 100) of histogram bins:");
    printf("%s", "> ");
    scanf("%li", &nt);
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* generate events and build histogram */
    rnbw(&c_d0, &c_d1, random, &nev);
    delt = 6. / (double) nt;
    smhsin_(hist, &c_dm3, &delt, &nt);
    i__1 = nev;
    for (i = 1; i <= i__1; ++i) {
        smhsfl_(hist, &c_dm3, &delt, &nt, &random[i - 1], &c_d1);
    }
    n = 0;
    i__1 = nt;
    for (i = 1; i <= i__1; ++i) {
        if (hist[i - 1] > 0.) {
            ++n;
            t[n - 1] = ((double) i - .5) * delt - 3.;
            y[n - 1] = hist[i - 1];
            deltay[n - 1] = sqrt(y[n - 1]);
        }
    }
    if (n <= nr) {
        printf("%s%li%s\n", "only ", n ," bins with at least one event");
        exit(0);
    }
/* set first approximation of unknowns */
    x[0] = .5;
    x[1] = .5;
    x[2] = (double) nev;
/* perform fit */
    nstep = 100;
    lsqnon_(bwfnct, t, y, deltay, &n, &nr, &nr, list, x, cx, &r, a,
           scrat, &nstep);
    i__1 = n - nr;
    p = 1. - scchi2_(&r, &i__1);
    if (nstep < 0) {
        printf("%s%5li\n", "LSQNON ended with NSTEP = ", nstep);
    } else {
/* prepare graphics */
        memset(tx, ' ', 75);
        memset(ty, ' ', 75);
        strncpy(tx, "t", 1);
        strncpy(ty, "y", 1);
        ltx = 1;
        lty = 1;
        sprintf(capt, "x_1#=%5.2lf,x_2#=%5.2lf,x_3#=%5.1lf,"
                "&D@x_1#=%5.2lf,&D@x_2#=%5.2lf,&D@x_3#=%5.1lf",
                x[0], x[1], x[2], sqrt(cx[0]), sqrt(cx[4]), sqrt(cx[8]));
        lcapt = strlen(capt);
/* curve corresponding to solution */
        npl = 100;
        i__1 = npl;
        for (ipl = 1; ipl <= i__1; ++ipl) {
            xpl[ipl - 1] = (double) ipl * .08 - 4.;
            ypl[ipl - 1] = bwfnct(x, &nr, &xpl[ipl - 1]);
        }
/* graphical output */
        grhscv_(xpl, ypl, &npl, hist, &c_dm3, &delt, &nt, tx, &ltx, ty, &lty,
               capt, &lcapt, &nws);
    }
    return 0;
} /* main */

/* ------------------------------------------------------------------ */
void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                  double P_T *r, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */
    static integer i;

/* program generates random numbers corresponding to */
/* a Breit-Wigner distribution */
    /* Parameter adjustments */
    --r;

    /* Function Body */
    rnecuy_(&r[1], n);
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        r[i] = *a + *gamma * .5 * tan((r[i] - .5) * 3.14159);
    }
    return;
} /* rnbw */

/* ------------------------------------------------- */
double CBFUNC bwfnct(double const P_T *x, integer const P_T *nr,
                     double const P_T *t)
{
    /* System generated locals */
    double ret_val, d__1, d__2, d__3;

    /* Parameter adjustments */
    --x;

    /* Function Body */
/* Computing 2nd power */
    d__1 = x[2];
/* Computing 2nd power */
    d__2 = *t - x[1];
/* Computing 2nd power */
    d__3 = x[2];
    ret_val = x[3] * 2. * (d__1 * d__1) / (x[2] * 3.14159f * (d__2 * d__2 *
            4. + d__3 * d__3));
    return ret_val;
} /* bwfnct */

